services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,630 @@
+import io.vertx.ext.mongo.MongoClient;
+import io.vertx.ext.mongo.MongoClientDeleteResult;
+
+/**
+ * An implementation of the tenant service and the tenant management service
+ * that uses a mongodb database to store tenants.
+ * <p>
+ * On startup this adapter tries to find the tenant collection, if not found, it gets created.
+ */
+@SuppressWarnings("RedundantTypeArguments")
+@Component
+@Qualifier("serviceImpl")
+@ConditionalOnProperty(name = "hono.app.type", havingValue = "mongodb")
+public final class MongoDbBasedTenantService extends AbstractVerticle implements TenantService, TenantManagementService {
+
+    private static final Logger log = LoggerFactory.getLogger(MongoDbBasedTenantService.class);
+
+    private MongoDbCallExecutor mongoDbCallExecutor;
+    private MongoClient mongoClient;
+    private MongoDbBasedTenantsConfigProperties config;
+
+    /**
+     * Sets an the {@code MongoDbCallExecutor} and {@code MongoClient} for the mongodb connection.
+     *
+     * @param mongoDbCallExecutor An instance of the configured, connection ready mongodb executor singleton.
+     * @throws NullPointerException if the mongoDbCallExecutor is {@code null}.
+     */
+    @Autowired
+    public void setExecutor(final MongoDbCallExecutor mongoDbCallExecutor) {
+        this.mongoDbCallExecutor = Objects.requireNonNull(mongoDbCallExecutor);
+        this.mongoClient = Objects.requireNonNull(this.mongoDbCallExecutor.getMongoClient());
+    }
+
+    public MongoDbBasedTenantsConfigProperties getConfig() {
+        return config;
+    }
+
+    /**
+     * Sets the tenant specific configuration.
+     *
+     * @param configuration The tenant configuration
+     * @throws NullPointerException if the configuration is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final MongoDbBasedTenantsConfigProperties configuration) {
+        this.config = Objects.requireNonNull(configuration);
+    }
+
+    /**
+     * Starts the service.
+     * <p>
+     * Finishes if a tenant collection is found and creates an indexed collection else.
+     * </P>
+     *
+     * @param startPromise the startup promise
+     */
+    @Override
+    public void start(final Promise<Void> startPromise) {
+        final Promise<List<String>> existingCollections = Promise.promise();
+        mongoClient.getCollections(existingCollections);
+        existingCollections.future()
+                .compose(successExistingCollections -> {
+                    if (successExistingCollections.contains(getConfig().getCollectionName())) {
+                        return Future.succeededFuture();
+                    } else {
+                        // create index & implicit collection
+                        return mongoDbCallExecutor.createCollectionIndex(getConfig().getCollectionName(),
+                                new JsonObject().put(TenantConstants.FIELD_PAYLOAD_TENANT_ID, 1).put(TenantConstants.FIELD_PAYLOAD_DEVICE_ID, 1),
```

Comment:
Line too long. Please format it with the code formatter shipped with Hono. It is also applicable to several other places in this PR.